File: language/src/main/java/robok/lang/variables/VariableObject.java
package robok.lang.variables;

/*
  Essa classe é responsável por obter :
    model: Model da variável (Classe ou primitiva)
    type: Tipo da variável.
    name: Nome da variável.
    value: Valor da variável.
    
 This class is responsible for obtaining:
    model: Model of the variable (Class or primitive)
    type: Type of the variable.
    name: Name of the variable.
    value: Value of the variable.
 */    
public class VariableObject {
    
    private String model, modify_access = "default", type, name, value;
    private boolean is_static, is_native, is_final, is_synchronized, is_volatile, is_transient, is_abstract, is_strictfp;

    public VariableObject(){}

    public VariableObject(String model, String type, String name, String value){
        this.model = model;
        this.type = type;
        this.name = name;
        this.value = value;
    }

    public VariableObject(String model, String modify_access, String type, String name, String value){
        this.model = model;
        this.modify_access = modify_access;
        this.type = type;
        this.name = name;
        this.value = value;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }

    /*
     definir modificadores de uma vez só
     set modifiers at once
     */ 
    public void setModifiers(boolean is_static, boolean is_final, boolean is_native, boolean is_synchronized, boolean is_volatile, boolean is_transient, boolean is_abstract, boolean is_strictfp) {
        this.is_static = is_static;
        this.is_final = is_final;
        this.is_native = is_native;
        this.is_synchronized = is_synchronized;
        this.is_volatile = is_volatile;
        this.is_transient = is_transient;
        this.is_abstract = is_abstract;
        this.is_strictfp = is_strictfp;
    }
    
    /*Definir tipo de variavel pelo valor (Usado caso tipo for var)
      Set variable type by value (used if type var)
    */
    public static String setVariableTypeFromValue(String value) {
        // Checks if it is an integer number
        try {
            Integer.parseInt(value);
            return "int";
        } catch (NumberFormatException e) {
            // Not an integer number
        }

        // Checks if it is a long number
        try {
            Long.parseLong(value);
            return "long";
        } catch (NumberFormatException e) {
            // Not a long number
        }
        
        
        // Checks if it is a float number (must end with 'f' or 'F')
        if (value.endsWith("f") || value.endsWith("F")) {
            try {
                Float.parseFloat(value);
                return "float";
            } catch (NumberFormatException e) {
                // Not a float number
            }
        }
        
        // Checks if it is a double number
        try {
            Double.parseDouble(value);
            return "double";
        } catch (NumberFormatException e) {
            // Not a double number
        }

        // Checks if it is a boolean
        String lowerCaseValue = value.toLowerCase();
        if ("true".equals(lowerCaseValue) || "false".equals(lowerCaseValue)) {
            return "boolean";
        }

        // Checks if it is a single character
       /* if (value.length() == 1) {
            return "char";
        }/*
        
      /*thdev: failure to identify find, the value ends up arriving 
        with variable signs ' ' or " "
         
        I'll fix that part soon
     */

        // Otherwise, consider it as a String
        return "String";
    }

    /* 
     Métodos getters para verificar os valores dos modificadores
     Getter methods to check modifier values
     */
    public boolean isStatic() { return is_static; }
    public boolean isNative() { return is_native; }
    public boolean isFinal() { return is_final; }
    public boolean isSynchronized() { return is_synchronized; }
    public boolean isVolatile() { return is_volatile; }
    public boolean isTransient() { return is_transient; }
    public boolean isAbstract() { return is_abstract; }
    public boolean isStrictfp() { return is_strictfp; }
}

File: robok/robok-lang/src/main/java/org/robok/lang/variable/VariableObject.kt
package org.robok.lang.variable

/*
 *  This file is part of Robok © 2024.
 *
 *  Robok is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Robok is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with Robok.  If not, see <https://www.gnu.org/licenses/>.
 */

/*
  This class is responsible for obtaining:
    model: Model of the variable (Class or primitive)
    type: Type of the variable.
    name: Name of the variable.
    value: Value of the variable.
 */

class VariableObject(
    var model: String = "",
    var modifyAccess: String = "default",
    var type: String = "",
    var name: String = "",
    var value: String = ""
) {

    var isStatic: Boolean = false
    var isNative: Boolean = false
    var isFinal: Boolean = false
    var isSynchronized: Boolean = false
    var isVolatile: Boolean = false
    var isTransient: Boolean = false
    var isAbstract: Boolean = false
    var isStrictfp: Boolean = false

    // Secondary constructor
    constructor(model: String, type: String, name: String, value: String) : this(model, "default", type, name, value)

    // Set modifiers at once
    fun setModifiers(
        isStatic: Boolean,
        isFinal: Boolean,
        isNative: Boolean,
        isSynchronized: Boolean,
        isVolatile: Boolean,
        isTransient: Boolean,
        isAbstract: Boolean,
        isStrictfp: Boolean
    ) {
        this.isStatic = isStatic
        this.isFinal = isFinal
        this.isNative = isNative
        this.isSynchronized = isSynchronized
        this.isVolatile = isVolatile
        this.isTransient = isTransient
        this.isAbstract = isAbstract
        this.isStrictfp = isStrictfp
    }

    // Set variable type by value (used if type is var)
    companion object {
        fun setVariableTypeFromValue(value: String): String {
            // Checks if it is an integer number
            try {
                value.toInt()
                return "int"
            } catch (e: NumberFormatException) {
                // Not an integer number
            }

            // Checks if it is a long number
            try {
                value.toLong()
                return "long"
            } catch (e: NumberFormatException) {
                // Not a long number
            }

            // Checks if it is a float number (must end with 'f' or 'F')
            if (value.endsWith("f", true)) {
                try {
                    value.toFloat()
                    return "float"
                } catch (e: NumberFormatException) {
                    // Not a float number
                }
            }

            // Checks if it is a double number
            try {
                value.toDouble()
                return "double"
            } catch (e: NumberFormatException) {
                // Not a double number
            }

            // Checks if it is a boolean
            val lowerCaseValue = value.lowercase()
            if (lowerCaseValue == "true" || lowerCaseValue == "false") {
                return "boolean"
            }

            // Future fix for handling char detection
            // For now, consider it as a String
            return "String"
        }
    }
}